DirectEmployers/MyJobs
# [PD-1383] New Datasources

This PR creates working reports for the Partners and Communication Types report types.

To test:

    manage.py reset_dynamic_data
    cd gulp/
    npm install
    gulp watch-tasks

Go to http://.../reports/view/dynamicoverview

Working:

* There should be one reporting type: PRM
* All three report types: Comm Records, Contacts, and Partners should work.
* All filter fields in the report types should work and provide meaningful help.

Not working:

* The report pages are still hideous and do not conform with our standards.

Review Guide:

* The bulk of this PR is the addition of the new partners and communication records report types.
* Moved the datasources to their own directory since each one is fairly large.
* There is some duplication between the datasource classes but the methods are subtly different between them. (IMHO further factoring would lead to very abstract code and inflexibility when creating a datasource for data that is not a Django model.)
* I did factor out a common DataSourceJsonDriver. It is responsible for adapting the datasources to the api views and was extremely similar.

- [x] Forward "Usability Nits" to @fezick for new tickets.
- [x] Replace `get_datasource_json_driver` with a plain old dictionary.
- [x] Use `abc` or otherwise document the datasource protocol.
- [x] Don't shadow `filter`.
- [x] Rename `$abbrev_qs` vars to be more descriptive.
- [x] Eliminate need for `'pass'` in `filter_key_types`.
- [x] Include `filter_key_types` in documentation.
- [x] Consider `unicodecsv` over `csv`.
- [x] pep8
- [x] download report: 'NoneType' object has no attribute 'strftime''
- [x] Clarify unit test docs.
- [x] Detect unicode in JoinFormatter fallback.
- [x] Expose tag `hex_color` for formatting later.
